MapboxCoreNavigationTests/RouteControllerTests.swift

```diff
@@ -1,116 +1,246 @@
     }
-    
+
     func testLocationShouldUseHeading() {
-        let navigation = setup.routeController
-        let firstLocation = setup.firstLocation
+        let navigation = dependencies.routeController
+        let firstLocation = dependencies.firstLocation
         navigation.locationManager(navigation.locationManager, didUpdateLocations: [firstLocation])
-        
+
         XCTAssertEqual(navigation.location!.course, firstLocation.course, "Course should be using course")
-        
+
         let invalidCourseLocation = CLLocation(coordinate: firstLocation.coordinate, altitude: firstLocation.altitude, horizontalAccuracy: firstLocation.horizontalAccuracy, verticalAccuracy: firstLocation.verticalAccuracy, course: -1, speed: firstLocation.speed, timestamp: firstLocation.timestamp)
-        
+
         let heading = CLHeading(heading: mbTestHeading, accuracy: 1)!
-        
+
         navigation.locationManager(navigation.locationManager, didUpdateLocations: [invalidCourseLocation])
         navigation.locationManager(navigation.locationManager, didUpdateHeading: heading)
-        
+
         XCTAssertEqual(navigation.location!.course, mbTestHeading, "Course should be using bearing")
     }
+
+    // MARK: - Events & Delegation
+
+    func testTurnstileEventSentUponInitialization() {
+        // MARK: it sends a turnstile event upon initialization
+        let _ = RouteController(along: initialRoute, directions: directionsClientSpy, locationManager: NavigationLocationManager(), eventsManager: eventsManagerSpy)
+
+        XCTAssertTrue(eventsManagerSpy.hasFlushedEvent(with: MMEEventTypeAppUserTurnstile))
+    }
+
+    func testReroutingFromALocationSendsEvents() {
+        let routeController = dependencies.routeController
+        let testLocation = dependencies.firstLocation
+
+        routeController.delaysEventFlushing = false
+
+        let willRerouteNotificationExpectation = expectation(forNotification: .routeControllerWillReroute, object: routeController) { (notification) -> Bool in
+            let fromLocation = notification.userInfo![RouteControllerNotificationUserInfoKey.locationKey] as? CLLocation
+            return fromLocation == testLocation
+        }
+
+        let didRerouteNotificationExpectation = expectation(forNotification: .routeControllerDidReroute, object: routeController, handler: nil)
+
+        let routeProgressDidChangeNotificationExpectation = expectation(forNotification: .routeControllerProgressDidChange, object: routeController) { (notification) -> Bool in
+            let location = notification.userInfo![RouteControllerNotificationUserInfoKey.locationKey] as? CLLocation
+            let rawLocation = notification.userInfo![RouteControllerNotificationUserInfoKey.rawLocationKey] as? CLLocation
+            let _ = notification.userInfo![RouteControllerNotificationUserInfoKey.routeProgressKey] as! RouteProgress
```

Comment:
Why are we pulling this out if we're just throwing it away?

Reply:
It would be more explicit to nil check it, agreed.